File: leetcode/143-reorder-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */
var reorderList = function(head) {
  if (!head) return;

  var slow, fast;
  for (slow = head, fast = head; fast.next && fast.next.next;
       slow = slow.next, fast = fast.next.next) ;

  var tail = null;
  for (var cur = slow.next; cur; ) {
    var next = cur.next;
    cur.next = tail;
    [tail, cur] = [cur, next];
  }

  slow.next = null;
  for (var i = head, j = tail; i && j; ) {
    var a = i.next, b = j.next;
    [i.next, j.next] = [j, a];
    [i, j] = [a, b];
  }
};

File: leetcode/143-reorder-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reorderList(self, head):
        """
        :type head: ListNode
        :rtype: void Do not return anything, modify head in-place instead.
        """
        if head is None:
            return

        slow, fast = head, head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next

        tail = None
        cur = slow.next
        while cur:
            nex = cur.next
            cur.next = tail
            tail, cur = cur, nex

        slow.next = None
        i, j = head, tail
        while i and j:
            a, b = i.next, j.next
            i.next, j.next = j, a
            i, j = a, b
